data-prepper-api/src/main/java/org/opensearch/dataprepper/model/source/SourceCoordinator.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.model.source;
+
+import java.time.Duration;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Interface to be passed to {@link org.opensearch.dataprepper.model.source.Source} to enable distribution and coordination of work
+ * between source plugins, and to track progress and save state for these partitions of work
+ * @since 2.2
+ */
+public interface SourceCoordinator {
+
+    /**
+     * A partition will be created for each of the partitionKeys passed.
+     * @param partitionKeys
+     */
+    void createPartitions(final List<String> partitionKeys);
```

Comment:
Should these partitions have some sort of ordering?

For reading from S3 this probably is the string provided here. But, for an OpenSearch index, we might want to provide the `creation_date`.

Maybe this should be provided as an additional string?

```
class Partition {
  String key;
  String orderKey;
}

...

createPartitions(final<Partition> partitions);
```

In this way, we could list indexes in OpenSearch and we know the latest created index.